ts/session/protocols/SessionProtocol.ts

```diff
@@ -1,56 +1,314 @@
+  public static async hasSession(device: string): Promise<boolean> {
+    // Session does not use the concept of a deviceId, thus it's always 1
+    const address = new libsignal.SignalProtocolAddress(device, 1);
+    const sessionCipher = new libsignal.SessionCipher(
+      textsecure.storage.protocol,
+      address
+    );
 
-export function shouldProcessSessionRequest(
-  device: string,
-  messageTimestamp: number
-): boolean {
-  // TODO: Need to do the following here
-  // messageTimestamp > session request sent timestamp && messageTimestamp > session request processed timestamp
-  return false;
-}
+    return sessionCipher.hasOpenSession();
+  }
+
+  /**
+   * Returns true if we sent a session request to that device already OR
+   *  if a session request to that device is right now being sent.
+   */
+  public static async hasSentSessionRequest(device: string): Promise<boolean> {
+    const pendingSend = SessionProtocol.pendingSendSessionsTimestamp.has(
+      device
+    );
+    const hasSent = await SessionProtocol._hasSentSessionRequest(device);
+
+    return pendingSend || hasSent;
+  }
+
+  /**
+   * Triggers a SessionResetMessage to be sent if:
+   *   - we do not already have a session and
+   *   - we did not sent a session request already to that device and
+   *   - we do not have a session request currently being send to that device
+   */
+  public static async sendSessionRequestIfNeeded(
+    device: string
+  ): Promise<void> {
+    if (
+      (await SessionProtocol.hasSession(device)) ||
+      (await SessionProtocol.hasSentSessionRequest(device))
+    ) {
+      return Promise.resolve();
+    }
+
+    const preKeyBundle = await libloki.storage.getPreKeyBundleForContact(
+      device
+    );
+
+    const sessionReset = new SessionResetMessage({
+      preKeyBundle,
+      timestamp: Date.now(),
+    });
+
+    return SessionProtocol.sendSessionRequest(sessionReset, device);
+  }
+
+  /**  */
+  public static async sendSessionRequest(
+    message: SessionResetMessage,
+    device: string
+  ): Promise<void> {
+    const timestamp = Date.now();
+
+    // mark the session as being pending send with current timestamp
+    // so we know we already triggered a new session with that device
+    SessionProtocol.pendingSendSessionsTimestamp.add(device);
+
+    // FIXME to remove
+    function toRawMessage(m: any): RawMessage {
+      return {
+        identifier: 'identifier',
+        plainTextBuffer: new TextEncoder().encode('jk'),
+        timestamp: Date.now(),
+        device: 'device',
+        ttl: 10,
+        encryption: EncryptionType.SessionReset,
+      };
+    }
+
+    try {
+      // TODO: Send out the request via MessageSender
+      const rawMessage = toRawMessage(message);
+      await MessageSender.send(rawMessage);
+      await SessionProtocol.updateSentSessionTimestamp(device, timestamp);
+    } catch (e) {
+      // console.log('Failed to send session request to:', device);
```

Comment:
i think this should throw the error, while we should catch in `sendSessionRequestIfNeeded` and log there.

This way if we want to send directly we can determine wether the send succeeded or failed, where as with `sendSessionRequestIfNeeded` we wouldn't really care if it sent or not.